aerleon/aerleon
# YAML schemas

This PR completes work on the YAML schemas. Schemas are provided for config files (e.g. aerleon.yml), definition files (e.g. naming.yaml) and policy files.

These schemas have been validated using the AJV validator through schemastore's test suite:

```
$ npm run remote

> schemastore.org@1.0.0 remote
> grunt remote_test

Running "remote_assert_schema_no_bom" task

Running "remote_test_ajv" task

>> pass schema          | http://172.23.0.3:8000/schemas/aerleon-config.schema.json (draft-04)(FullStrictMode)
>> http://172.23.0.3:8000/schemas/aerleon-config.schema.json

>> pass schema          | http://172.23.0.3:8000/schemas/aerleon-definitions.schema.json (draft-07)(FullStrictMode)
>> http://172.23.0.3:8000/schemas/aerleon-definitions.schema.json

>> pass schema          | http://172.23.0.3:8000/schemas/aerleon-policies.schema.json (draft-07)(FullStrictMode)
>> http://172.23.0.3:8000/schemas/aerleon-policies.schema.json

Total schemas validated with AJV: 3
```

To test these changes in VSCode, perform the following steps. Note that once the schema is provided to json-schema-store, the RedHat YAML extension will no longer be required to use these schemas.

* Check out this branch.
* Install the RedHat YAML Extension: https://marketplace.visualstudio.com/items?itemName=redhat.vscode-yaml
* Open the Settings panel in VSCode, search for "Schema", and edit your settings.json with the following config:
```
    "yaml.schemas": {
    
        "schemas/aerleon-config.schema.json": ["aerleon.yml"],
        "schemas/aerleon-policies.schema.json": ["*.policy.yaml", "policies/**/*.yaml"],
        "schemas/aerleon-definitions.schema.json": ["*.definition.yaml", "def/**/*.yaml"],
    },
```
* Open the aerleon.yml file to test the config schema.
* Open any policy YAML file to test the policy schema.
* Open the naming.yaml file to test the definitions schema.

**Future Work**

This PR does not attempt to document or validate any specific term keyword. A future version of this schema could provide detailed text for each built-in term keyword. Validating each term value may not be a good idea though, as we want to reserve the right for generator authors to extend the configuration language.
